src/runtime/runtime.ts

```diff
@@ -1649,61 +1650,69 @@ export class JoinNode extends Node {
   }
 
   exec(context:EvaluationContext, input:Change, prefix:Prefix, transaction:number, round:number, results:Iterator<Prefix>):boolean {
-    if(this.isStatic && this.dormant) return false;
-
-    this.inputCount = input.count;
     let didSomething = false;
-    let affectedConstraints = this.findAffectedConstraints(input, prefix);
-
-    let combinationCount = Math.pow(2, affectedConstraints.length);
-    for(let comboIx = combinationCount - 1; comboIx > 0; comboIx--) {
-      //console.log("  Combo:", comboIx);
-
-      let shouldApply = true;
-
-      for(let constraintIx = 0; constraintIx < affectedConstraints.length; constraintIx++) {
-        let mask = 1 << constraintIx;
-        let isIncluded = (comboIx & mask) !== 0;
-        let constraint = affectedConstraints.array[constraintIx];
-        constraint.isInput = isIncluded;
-
-        if(isIncluded) {
-          let valid = constraint.applyInput(input, prefix);
-          // debug("        included", printConstraint(constraint));
-          // If any member of the input constraints fails, this whole combination is doomed.
-          if(valid === ApplyInputState.fail) {
-            shouldApply = false;
-            break;
+    this.inputCount = input.count;
+
+    if(input === BLOCK_REMOVE) {
```

Comment:
We should add a comment for add and remove like "to remove a block we run with no applied inputs and a negative count, which gives us retractions of all the changes the block produces."